oauth2_http/java/com/google/auth/oauth2/S2A.java

```diff
@@ -0,0 +1,107 @@
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ServiceLoader;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utilities to fetch the S2A (Secure Session Agent) address from the mTLS configuration.
+ *
+ * <p>mTLS configuration is queried from the MDS MTLS Autoconfiguration endpoint.
+ */
+@ThreadSafe
+public final class S2A {
+  public static final String DEFAULT_METADATA_SERVER_URL = "http://169.254.169.254";
+  public static final String MTLS_CONFIG_ENDPOINT =
+      "/computeMetadata/v1/instance/platform-security/auto-mtls-configuration";
+
+  public static final String METADATA_FLAVOR = "Metadata-Flavor";
+  public static final String GOOGLE = "Google";
+  private static final String PARSE_ERROR_S2A = "Error parsing Mtls Auto Config response.";
+
+  private MtlsConfig config;
+
+  private transient HttpTransportFactory transportFactory;
+
+  public S2A() {}
+
+  public void setHttpTransportFactory(HttpTransportFactory tf) {
+    this.transportFactory = tf;
+  }
+
+  /** @return the mTLS S2A Address from the mTLS config. */
+  public synchronized String getMtlsS2AAddress() {
+    if (config == null) {
+      config = getMdsMtlsConfig();
+    }
+    return config.getMtlsS2AAddress();
+  }
+
+  /** @return the plaintext S2A Address from the mTLS config. */
+  public synchronized String getPlaintextS2AAddress() {
+    if (config == null) {
+      config = getMdsMtlsConfig();
+    }
+    return config.getPlaintextS2AAddress();
+  }
+
+  /**
+   * Queries the MDS mTLS Autoconfiguration endpoint and returns the {@link MtlsConfig}.
+   *
+   * <p>Returns {@link MtlsConfig} with empty addresses on error.
+   *
+   * @return the {@link MtlsConfig}.
+   */
+  private MtlsConfig getMdsMtlsConfig() {
+    String plaintextS2AAddress = "";
+    String mtlsS2AAddress = "";
+    try {
+      if (transportFactory == null) {
+        transportFactory =
+            Iterables.getFirst(
+                ServiceLoader.load(HttpTransportFactory.class), OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+      }
+      String url = getMdsMtlsEndpoint();
+      GenericUrl genericUrl = new GenericUrl(url);
+      HttpRequest request =
+          transportFactory.create().createRequestFactory().buildGetRequest(genericUrl);
+      JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+      request.setParser(parser);
+      request.getHeaders().set(METADATA_FLAVOR, GOOGLE);
+      request.setThrowExceptionOnExecuteError(false);
+      HttpResponse response = request.execute();
+
+      if (!response.isSuccessStatusCode()) {
+        return MtlsConfig.createBuilder().build();
+      }
+
+      InputStream content = response.getContent();
+      if (content == null) {
+        return MtlsConfig.createBuilder().build();
+      }
```

Comment:
is it possible to reuse the code below  for querying mds endpoint?
https://github.com/googleapis/google-auth-library-java/blob/dde6876154d6e7ff43dbbf7b75c5205eadd679a5/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L338

Reply:
We could use this function to get the MDS HttpResponse, but it would only replace a few lines (~7) in this function: 

creating the HttpRequest and executing it to get the response, replaced with

```response = getMetadataResponse(url)```.

However, in order to do this, we would also have to:
- ignore the ```ComputeEngineCredentials``` specific errors thrown by the function, because we only care if we are able to successfully create a response (aligning with Go implementation, return empty S2A Address if any error). This technically works, although I am not sure it is best practice.
- ```getMetadataResponse(String url)``` is not static, so we would have to create an instance of ```ComputeEngineCredentials``` to use it.

WDYT?